### PR TITLE
Add packages by tracking number

### LINKED ISSUE #53: Feature: add tracking number
Is there any reason why adding a tracking number to the user's profile is not implemented yet? I think it would be cool to have a Home Assistant service for that.

I looked at the API calls in the background, and adding a tracking number calls `https://buyer.17track.net/orderapi/call` with this payload:

```json
{
    "version": "1.0",
    "timeZoneOffset": -60,
    "method": "AddTrackNo",
    "param": {
        "TrackNos": [
            "123456789"
        ]
    }
}
```

If this feature makes sense, I would be happy to open a PR.

### PR DESCRIPTION
**Pull request by ofalvai**
**Describe what the PR does:**

Adds a new method (`add_packages`) to add one or more packages to track by their tracking number.

Unfortunately, I couldn't find a way to add a friendly name to each new package. The API method that adds a friendly name (`SetTrackRemark`) expects an internal ID, not the tracking number itself. This internal ID isn't returned when adding a new package, only when getting all tracked packages.

@bachya what do you think, would it make sense to create a method that calls `packages()` first to get that internal ID, and then sets a friendly name for that ID?

**Does this fix a specific issue?**

Fixes https://github.com/bachya/py17track/issues/53
  
**Checklist:**

- [X] Confirm that one or more new tests are written for the new functionality.
- [X] Update `README.md` with any new documentation.
- [X] Add yourself to `AUTHORS.md`.

### CODE REVIEW

**Comment by bachya** [comment]
>@bachya what do you think, would it make sense to create a method that calls packages() first to get that internal ID, and then sets a friendly name for that ID?

This isn't the worst thing, but we should only do it if the user provides a friendly name – perhaps a `friendly_name` kwarg and if it's provided, we do the flow you suggest?

**Comment by ofalvai** [comment]
@bachya Done, please take a look :)

File: py17track/profile.py
Comment (bachya) [inline py17track/profile.py:80]: You define `id` as a required attribute of the `Package` class, but this implies that the value could be `None`. Is that expected? Either way is ultimately fine, but we should be consistent (i.e., if it's expected, access it directly; if it can be `None`, make it an optional attribute).

File: py17track/profile.py
Comment (bachya) [inline py17track/profile.py]: We shouldn't swallow an unsuccessful response here. If we don't get an `HTTP 2xx` response, we should let the `RequestError` already in `self._request` bubble up. If we could get an `HTTP 2xx` response with a non-0 `Code`, we should still raise a `RequestError`.

Comment (bachya) [reply py17track/profile.py]: This would eliminate the need for returning a `bool`, as well – if an exception isn't raised, we can assume the request was successful.

File: py17track/profile.py
Comment (bachya) [inline py17track/profile.py]: We can avoid nesting here by reversing things into a guard clause:

```python
if not friendly_name:
  return
```

File: py17track/profile.py
Comment (bachya) [inline py17track/profile.py]: Using filter with a lambda feels very un-Pythonic. How about:

```python
try:
  new_package = next(p for p in packages if p.tracking_number == tracking_number)
except StopIteration:
  # Raise an exception? Return?
```

File: py17track/profile.py
Comment (bachya) [inline py17track/profile.py]: Reverse this to avoid deep nesting.

File: py17track/profile.py
Comment (bachya) [inline py17track/profile.py]: Don't swallow errors.

File: py17track/profile.py
Comment (bachya) [inline py17track/profile.py]: Don't swallow errors.

**Review by bachya — CHANGES_REQUESTED** [review]

File: py17track/profile.py
Comment (ofalvai) [reply py17track/profile.py:80]: Good catch, I chose the optional ID.

**Comment by ofalvai** [comment]
Thanks @bachya for the detailed review comments. I tried to fix all of them, but let me know if something is still not right (as you might have guessed, I'm not that familiar with the Python world 😁).
I also leave it to you to decide the fate of that cyclomatic complexity warning.

File: py17track/package.py
Comment (bachya) [inline py17track/package.py]: Small nit: let's group the `Optional` attributes together.

**Comment by bachya** [comment]
Looking good overall! One little comment, plus you need to make sure you write tests to cover new functionality. Don't worry about the Code Climate warning; I'm not concerned.

**Comment by ofalvai** [comment]
Done, code coverage is 100% again :)

**Review by bachya — APPROVED** [review]
Good to merge once CI passes!

**Pull request merged**
